src/Mono.Android/Java.Interop/TypeManager.cs

```diff
@@ -243,15 +243,26 @@ static Exception CreateJavaLocationException ()
 			Type? type = null;
 			IntPtr class_ptr = JNIEnv.GetObjectClass (handle);
 			string? class_name = GetClassName (class_ptr);
+			List<string>? class_names = null;
 			lock (TypeManagerMapDictionaries.AccessLock) {
 				while (class_ptr != IntPtr.Zero && !TypeManagerMapDictionaries.JniToManaged.TryGetValue (class_name, out type)) {
-
 					type = GetJavaToManagedType (class_name);
 					if (type != null) {
 						TypeManagerMapDictionaries.JniToManaged.Add (class_name, type);
+
+						// cache shortcuts for the class names of the subclasses in the hierarchy we just explored
+						if (class_names is not null) {
+							foreach (var subclass_name in class_names) {
+								TypeManagerMapDictionaries.JniToManaged.Add (subclass_name, type);
```

Comment:
Is the `type` here actually the correct one? It seems like it would be the final type it found and not the one that matches the `class_name` earlier in the loop.

Do you need to store a `(string, Type)` tuple, so you have the previous `Type` instance?

Reply:
You're right that it is not an exact match. On the other hand, by inspecting the superclasses in the type hierarchy we should always arraive at the same `type`. So my idea was to cache this "shortcut".

The only problem I can think of is if we stored this "shortcut", any later call to `TypeManager.RegisterType` with a more specific mapping would be ignored. This could break dynamic registration. I will need to give this some more thought.